### ISSUE TITLE
Feature Request: Start optimization from given model

### ISSUE THREAD

**Issue by mmjb** [Z3Prover/z3#753]
It would be nice to have a z3-opt API like `optimize(model_ref& base_model)`, which starts optimisation from the given base model. Currently, `optimize()` always calls into a fresh solver to get a solution on the hard constraints before starting the optimisation procedure, which can be wasteful in some situations.

**Comment by NikolajBjorner** [comment]
It is unclear how this could be leveraged: internally the formulas asserted to the solvers may have been rewritten and be using fresh variables.
